src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+        header then calls the appropriate smb_XXX
+        method with data beyond the header.
+
+        @param packet: the raw packet
+        @type packet: L{bytes}
+        """
+        protocol_id = packet[:4]
+        if protocol_id == b"\xFFSMB":
+            # its a SMB1 packet which we dont support with the exception
+            # of the first packet, we try to offer upgrade to SMB2
+            if self.avatar is None:
+                log.debug("responding to SMB1 packet")
+                self.negotiate_response()
+            else:
+                self.transport.close()
+                log.error("Got SMB1 packet while logged in")
+            return
+        elif protocol_id != b"\xFESMB":
+            self.transport.close()
+            log.error("Unknown packet type")
+            log.debug("packet data {data!r}", data=packet[:64])
+            return
+        begin_struct = "<4xHH4sHHLLQ"
+        (hdr_size, self.credit_charge, hdr_status, self.hdr_command,
+         self.credit_request, self.hdr_flags, self.next_command,
+         self.message_id) = struct.unpack(begin_struct, packet[:32])
+        self.is_async = (self.hdr_flags & FLAG_ASYNC) > 0
+        self.is_related = (self.hdr_flags & FLAG_RELATED) > 0
+        self.is_signed = (self.hdr_flags & FLAG_SIGNED) > 0
+        # FIXME other flags 3.1 or too obscure
+        if self.is_async:
+            (self.async_id, self.session_id,
+             self.signature) = struct.unpack("<QQ16s", packet[32:64])
+            self.tree_id = 0x00
+        else:
+            (_reserved, self.tree_id, self.session_id,
+             self.signature) = struct.unpack("<LLQ16s", packet[32:64])
+            self.async_id = 0x00
+        log.debug("HEADER")
+        log.debug("------")
+        log.debug("protocol ID     {pid!r}", pid=protocol_id)
+        log.debug("size            {hs}", hs=hdr_size)
+        log.debug("credit charge   {cc}", cc=self.credit_charge)
+        log.debug("status          {status}", status=hdr_status)
+        log.debug("command         {cmd!r} {cmdn:02x}",
+                  cmd=COMMANDS[self.hdr_command],
+                  cmdn=self.hdr_command)
+        log.debug("credit request  {cr}", cr=self.credit_request)
+        s = ""
+        if self.is_async:
+            s += "ASYNC "
+        if self.is_signed:
+            s += "SIGNED "
+        if self.is_related:
+            s += "RELATED "
+        log.debug("flags           0x{flags:x} {s}", flags=self.hdr_flags, s=s)
+        log.debug("next command    0x{nc:x}", nc=self.next_command)
+        log.debug("message ID      0x{mid:x}", mid=self.message_id)
+        log.debug("session ID      0x{sid:x}", sid=self.session_id)
+        if self.is_async:
+            log.debug("async ID        0x{aid:x}", aid=self.async_id)
+        else:
+            log.debug("tree ID         0x{tid:x}", tid=self.tree_id)
+        log.debug("signature       {sig}",
+                  sig=binascii.hexlify(self.signature))
```

Comment:
This should really be one large log message, rather than tons of tiny individual ones.  Thinking from the perspective of a log *consumer*, this is a single event, and you wouldn't want to write code to try to heuristically figure out where it begins and ends.  We have had much bitter experience with the annoying split-up message in `defer.py` that does this, that should be a single message.